### PR TITLE
Cleanup ml prover 5

### PR DESCRIPTION
**Pull request by yilongli**
@dwightguth @andreistefanescu please review

### CODE REVIEW

File: java-backend/src/main/java/org/kframework/backend/java/kil/ConstrainedTerm.java [hunk 1]
```diff
@@ -136,16 +136,10 @@ public SymbolicConstraint matchImplies(ConstrainedTerm constrainedTerm) {
...
-        // TODO(AndreiS): figure out when a map unification provided more data
-        unificationConstraint.expandPatternsAndSimplify(false);
-        unificationConstraint.expandPatternsAndSimplify(false);
-        unificationConstraint.expandPatternsAndSimplify(false);
 
         final Set<Variable> variables = Sets.newHashSet(unificationConstraint.variableSet());
         variables.removeAll(variableSet());
-        unificationConstraint.orientSubstitution(variables);
-        if (unificationConstraint.isFalse()
-                || !unificationConstraint.substitution().keySet().containsAll(variables)) {
```
Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/kil/ConstrainedTerm.java:148]: Why did you remove this condition?

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java [hunk 2]
```diff
@@ -519,10 +519,17 @@ public ASTNode visit(org.kframework.kil.Rule node, Void _)  {
...
-        Collection<Variable> freshVariables = new ArrayList<>();
+        Set<Variable> freshConstants = Sets.newHashSet();
         // TODO(AndreiS): check !Variable only appears in the RHS
         for (org.kframework.kil.Variable variable : node.getBody().variables()) {
             if (variable.isFreshConstant()) {
+                freshConstants.add((Variable) this.visitNode(variable));
+            }
+        }
+
+        Set<Variable> freshVariables = Sets.newHashSet();
```
Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java]: you did not refactor this into a lambda :-)

**Comment by andreistefanescu** [comment]
Looks good. This refactoring and the previous one are really useful!

File: java-backend/src/main/java/org/kframework/backend/java/kil/ConstrainedTerm.java (hunk 1 above)
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/kil/ConstrainedTerm.java:148]: because `orientSubstitution` now returns a boolean to indicate whether the orientation succeeds (the orientation is possible and the resulting constraint is not false). If so, then there is no need to check `unificationConstraint.substitution().keySet().containsAll(variables)`.  If not, then the substitution will remain unchanged.

**Comment by dwightguth** [comment]
lgtm. You can merge if you are happy with this, Yilong.
